### ISSUE TITLE
[dynamo] autocast context manager is lowered without try/finally, so an exception leaks autocast state

### ISSUE THREAD

**Issue by bobrenjc93** [pytorch/pytorch#195488]
### Summary

Dynamo lowers `with autocast(...)` into three straight-line graph nodes with no `try`/`finally`, so a runtime exception between the enter and the exit permanently skips the exit:

```python
_enter_autocast = torch.amp.autocast_mode._enter_autocast('cpu', torch.bfloat16, True, None)
index_select = l_x_.index_select(0, l_idx_)                              # raises here
_exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast) # never runs
```

Eager's `with` restores all four pieces of autocast state on the way out. The compiled version does not.

### Repro

```python
import torch

def probe():
    n = torch.autocast_increment_nesting(); torch.autocast_decrement_nesting()
    return n - 1, torch.is_autocast_enabled('cpu'), torch.get_autocast_dtype('cpu')

@torch.compile(backend='eager', fullgraph=True)
def f(x, idx):
    with torch.amp.autocast('cpu', dtype=torch.bfloat16):
        return x.index_select(0, idx)

x = torch.randn(4, 4); bad = torch.tensor([99])
print('before', probe())
for i in range(3):
    try:
        f(x, bad)
    except IndexError:
        pass
    print('after', i, probe())
```

Output on clean `main`:

```
before (0, False, torch.bfloat16)
after 0 (1, False, torch.bfloat16)
after 1 (2, False, torch.bfloat16)
after 2 (3, False, torch.bfloat16)
```

The nesting counter walks 0 -> 1 -> 2 -> 3 and `fast_dtype` leaks with it. Only `enabled` comes back, and only because it is restored separately by the global-state checkpoint in `torch/_dynamo/output_graph.py`.

### Why it matters

A leaked nesting counter means a later `autocast.__exit__` never sees the counter reach 0, so `clear_autocast_cache()` never runs and a weight cached in one dtype survives into a region using another, surfacing as `mat1 and mat2 must have the same dtype`. That is a silent-wrong-numerics or confusing-error failure mode some distance from its cause. See #194754, where the same end symptom arose from a different mechanism (a test calling `autocast_decrement_nesting()` with no matching increment).

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @chauhang @penguinwu @voznesenskym @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98 @anijain2305 @williamwen42 @jansel

This issue was written with assistance from an AI coding assistant; the repro above was run by me on clean `main`.

**Comment by saiyambharara** [comment]
Hi [@bobrenjc93](https://github.com/bobrenjc93), I'd like to work on this and have done some initial testing. I checked whether the missing `try/finally` is specific to autocast or also affects other context managers that Dynamo lowers into enter/exit nodes. It looks like the same issue applies to several of them, so I wanted to check with you on the intended scope before putting up a PR.

I tested the context managers that emit both enter and exit nodes from Dynamo. For each one, I ran the same case and compared the state after an exception leaves the compiled region with the corresponding eager case:

```text
context manager                         verdict    eager                 compiled
----------------                         -------    -----                 --------
autocast('cpu')                          LEAKS      0                     3
no_grad()                                LEAKS      True                  False
inference_mode()                         ok         False                 False
_set_fwd_grad_enabled(False)             LEAKS      True                  False
disable_saved_tensors_hooks              LEAKS      None                  'probe'
sdpa_kernel(MATH)                        LEAKS      (True, True, True)    (False, True, False)
```

Each row reads whatever that manager's `__exit__` is responsible for restoring, hence the differing types:

- `autocast`: the nesting counter. This one accumulates, gaining +1 per skipped exit and growing without bound in a loop that catches exceptions. 

- `no_grad()` and `_set_fwd_grad_enabled`: `is_grad_enabled()` and `is_fwd_grad_enabled()`. Autograd or forward-mode AD stays off, so later training silently computes no gradients.
- `disable_saved_tensors_hooks`: the disabled-reason message, where `None` means hooks are working. Hooks stay disabled and later errors quote a stale reason from a region that already died.
- `sdpa_kernel`: the `(flash, math, mem_efficient)` flags. The global attention-backend selection stays overridden, so every later model in the process silently uses a different kernel with nothing pointing back at the cause.

`inference_mode` returns to baseline. There are more classes that share the emission shape, so this list is probably not exhaustive. Would you prefer that I broaden the issue to cover the general case, or keep this focused on autocast and track the other context managers separately? Also, is there a particular approach you'd prefer for the fix, or any thoughts on the direction I should take?

<sub>The sweep table and script above were produced with AI assistance. I have reviewed the code and reproduced the results myself.</sub>

**Comment by anijain2305** [comment]
This is ... hard.

Fx graphs are straight line code. We can do some tricks like `__enter__` registers an exit action, `__exit__` unregisters it, and then at the end compiled frame runs the outstanding actions. But in general, this is pretty hard to do.

**Comment by jansel** [comment]
Would need some sort of machinery in the FX graph.  Something like a single [ExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack) per execution.
